src/Orleans.Runtime/Streams/QueueBalancer/LeaseBasedQueueBalancer.cs

```diff
@@ -9,84 +9,25 @@
-    /// <summary>
-    /// IResourceSelector selects a certain amount of resources from a resource list
-    /// </summary>
-    /// <typeparam name="T"></typeparam>
-    internal interface IResourceSelector<T>
-    {
-        /// <summary>
-        /// Number of resources
-        /// </summary>
-        int Count { get; }
-
-        /// <summary>
-        /// Try to select certain count of resources from resource list, which doesn't overlap with existing selection
-        /// </summary>
-        /// <param name="newSelectionCount"></param>
-        /// <param name="existingSelection"></param>
-        /// <returns></returns>
-        List<T> NextSelection(int newSelectionCount, List<T> existingSelection);
-    }
-
-    /// <summary>
-    /// Selector using round robin algorithm
-    /// </summary>
-    /// <typeparam name="T"></typeparam>
-    internal class  RoundRobinSelector<T> : IResourceSelector<T>
-    {
-        private ReadOnlyCollection<T> resources;
-        private int lastSelection;
-        public RoundRobinSelector(IEnumerable<T> resources)
-        {
-            var rand = new Random(this.GetHashCode());
-            // distinct randomly ordered readonly collection
-            this.resources = new ReadOnlyCollection<T>(resources.Distinct().OrderBy(_ => rand.Next()).ToList());
-            this.lastSelection = rand.Next(this.resources.Count);
-        }
-
-        public int Count => this.resources.Count;
-
-        /// <summary>
-        /// Try to select certain count of resources from resource list, which doesn't overlap with existing resources
-        /// </summary>
-        /// <param name="newSelectionCount"></param>
-        /// <param name="existingSelection"></param>
-        /// <returns></returns>
-        public List<T> NextSelection(int newSelectionCount, List<T> existingSelection)
-        {
-            var selection = new List<T>(Math.Min(newSelectionCount, this.resources.Count));
-            int tries = 0;
-            while (selection.Count < newSelectionCount && tries++ < this.resources.Count)
-            {
-                this.lastSelection = (++this.lastSelection) % (this.resources.Count);
-                if(!existingSelection.Contains(this.resources[this.lastSelection]))
-                    selection.Add(this.resources[this.lastSelection]);
-            }
-            return selection;
-        }
-    }
-
     /// <summary>
     /// LeaseBasedQueueBalancer. This balancer supports queue balancing in cluster auto-scale scenarios,
     /// unexpected server failure scenarios, and tries to support ideal distribution as much as possible. 
     /// </summary>
     public class LeaseBasedQueueBalancer : QueueBalancerBase, IStreamQueueBalancer
     {
-        /// <summary>
-        /// Lease category for LeaseBasedQueueBalancer
-        /// </summary>
-        public const string LeaseCategory = "QueueBalancer";
-
         private class AcquiredQueue 
         {
+            public int LeaseId { get; set; }
```

Comment:
What does this id represent, is it an index? I prefer a string if there's no specific meaning to it (and it isn't used as a lease itself)

Reply:
Yeah, poor name.  It's used to track the order in which the leases were obtained in, so that this information is available to algorithms that select which leases to release should the provider need to shed leases.
Will try to make it more clear.

Reply:
Found bug in lease ordering.. bleh.. fixing..